contracts/QuestTerminalKey.sol

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.16;
+
+import '@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/token/ERC721/extensions/ERC721URIStorageUpgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/token/ERC721/extensions/ERC721EnumerableUpgradeable.sol';
+import {OwnableUpgradeable} from './OwnableUpgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import '@openzeppelin/contracts-upgradeable/interfaces/IERC2981Upgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol';
+
+contract QuestTerminalKey is
+    Initializable,
+    ERC721Upgradeable,
+    ERC721EnumerableUpgradeable,
+    ERC721URIStorageUpgradeable,
+    OwnableUpgradeable,
+    IERC2981Upgradeable
+{
+    event RoyaltyFeeSet(uint256 indexed royaltyFee);
+    event MinterAddressSet(address indexed minterAddress);
+    event QuestFactoryAddressSet(address indexed questFactoryAddress);
+    using CountersUpgradeable for CountersUpgradeable.Counter;
+    CountersUpgradeable.Counter private _tokenIds;
+
+    // storage
+    address public royaltyRecipient;
+    address public minterAddress;
+    address public questFactoryAddress;
+    uint public royaltyFee;
+    mapping(uint256 => Discount) public discounts;
+    struct Discount {
+        uint16 percentage; //in BIPS
+        uint16 maxUses;
+        uint16 usedCount;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        address royaltyRecipient_,
+        address minterAddress_,
+        address questFactoryAddress_,
+        uint royaltyFee_,
+        address owner_
+    ) external initializer {
+        __ERC721_init('QuestTerminalKey', 'QTK');
+        __ERC721URIStorage_init();
+        __Ownable_init(owner_);
+        royaltyRecipient = royaltyRecipient_;
+        minterAddress = minterAddress_;
+        questFactoryAddress = questFactoryAddress_;
+        royaltyFee = royaltyFee_;
+    }
+
+    modifier onlyMinter() {
+        require(msg.sender == minterAddress, 'Only minter');
+        _;
+    }
+
+    modifier onlyQuestFactory() {
+        require(msg.sender == questFactoryAddress, 'Only quest factory');
+        _;
+    }
+
+    function _baseURI() internal pure override returns (string memory) {
+        return "https://api.rabbithole.gg/nft/qtk/";
+    }
+
+    /// @dev modifier to check for zero address
+    /// @param _address the address to check
+    modifier nonZeroAddress(address _address) {
```

Comment:
Is there any benefit for this modifier? Does/can going without it save gas? (imo the check is more important than saving marginal amount of gas)